packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -695,136 +700,317 @@ class _SortData {
+  Iterable<FocusNode> sortDescendants(FocusScopeNode scope) {
+    final Iterable<FocusNode> geometryOrdered = secondary.sortDescendants(scope);
+    final List<FocusNode> unordered = <FocusNode>[];
+    final List<_OrderedFocusInfo> ordered = <_OrderedFocusInfo>[];
+    for (final FocusNode node in geometryOrdered) {
+      final FocusOrder order = FocusTraversalOrder.of(node.context, nullOk: true);
+      if (order != null) {
+        ordered.add(_OrderedFocusInfo(node: node, order: order));
+      } else {
+        unordered.add(node);
       }
     }
-    final List<FocusNode> sortedNodes = _sortByGeometry(nearestScope).toList();
-    if (forward && focusedChild == sortedNodes.last) {
-      _focusAndEnsureVisible(sortedNodes.first, alignmentPolicy: ScrollPositionAlignmentPolicy.keepVisibleAtEnd);
-      return true;
-    }
-    if (!forward && focusedChild == sortedNodes.first) {
-      _focusAndEnsureVisible(sortedNodes.last, alignmentPolicy: ScrollPositionAlignmentPolicy.keepVisibleAtStart);
-      return true;
-    }
+    ordered.sort((_OrderedFocusInfo a, _OrderedFocusInfo b) {
+      return a.order.compareTo(b.order);
+    });
+    return ordered.map<FocusNode>((_OrderedFocusInfo info) => info.node).followedBy(unordered);
+  }
+}
 
-    final Iterable<FocusNode> maybeFlipped = forward ? sortedNodes : sortedNodes.reversed;
-    FocusNode previousNode;
-    for (final FocusNode node in maybeFlipped) {
-      if (previousNode == focusedChild) {
-        _focusAndEnsureVisible(
-          node,
-          alignmentPolicy: forward
-            ? ScrollPositionAlignmentPolicy.keepVisibleAtEnd
-            : ScrollPositionAlignmentPolicy.keepVisibleAtStart,
+/// An inherited widget that describes the order in which its child subtree
+/// should be traversed.
+///
+/// The order for a widget is determined by the [FocusOrder] returned by
+/// [FocusTraversalOrder.of] for a particular context.
+class FocusTraversalOrder extends InheritedWidget {
+  /// A const constructor so that subclasses can be const.
+  const FocusTraversalOrder({Key key, this.order, Widget child}) : super(key: key, child: child);
+
+  /// The order for the widget descendants of this [FocusTraversalOrder].
+  final FocusOrder order;
+
+  /// Finds the nearest TraversalOrder widget.
+  ///
+  /// It does not create a rebuild dependency because changing the traversal
+  /// order doesn't change the widget tree until the traversal actually occurs,
+  /// so nothing needs to be rebuilt as a result of an order change.
+  static FocusOrder of(BuildContext context, {bool nullOk = false}) {
+    assert(context != null);
+    assert(nullOk != null);
+    final FocusTraversalOrder marker = context.findAncestorWidgetOfExactType<FocusTraversalOrder>();
+    final FocusOrder order = marker?.order;
+    if (order == null) {
+      if (!nullOk) {
+        throw FlutterError(
+            'TraversalOrder.of() was called with a context that does not contain a TraversalOrder widget.\n'
```

Comment:
FocusTraversalOrder.of and FocusTraversalOrder widget

Reply:
Same below.

Reply:
Done.